### PR TITLE
Cylinder repeat primitive

### PR DESCRIPTION
**Pull request by Clarkszw**
Fix the issue: Cylinder actor needs to use repeat_primitives by default #531
It has passed the test originally written for cylinder.
`pytest -svv fury/tests/test_actors.py::test_advanced_geometry_actor fury/tests/test_actors.py::test_actors_primitives_count`

Should I write another test for new `repeat_primitive` parameter in `cylinder()` ?

### CODE REVIEW

**Comment by skoudoro** [comment]
Thanks, Can you add a screenshot of the results, using primitives and not using primitives. Multiple Cylinders should be present with different scale and orientation.

Thanks!

File: fury/actor.py [hunk 1]
```diff
@@ -1883,18 +1882,21 @@ def cylinder(
     colors : ndarray (N,3) or (N, 4) or tuple (3,) or tuple (4,)
         RGB or RGBA (for opacity) R, G, B and A should be at the range [0, 1].
     radius : float
-        cylinder radius, default: 1.
+        cylinder radius, default: 0.05.
```
Comment (skoudoro) [inline fury/actor.py:1885]: you can remove the default information

File: fury/actor.py [hunk 2]
```diff
@@ -1883,18 +1882,21 @@ def cylinder(
...
     heights : ndarray, shape (N)
-        The height of the arrow.
+        The height of the cylinder, default: 1.
```
Comment (skoudoro) [inline fury/actor.py:1887]: same, remove `default`

**Review by skoudoro — COMMENTED** [review]
See below some additionnal comments. Thanks!

**Comment by Clarkszw** [comment]
> Thanks, Can you add a screenshot of the results, using primitives and not using primitives. Multiple Cylinders should be present with different scale and orientation.
> 
> Thanks!

Here is the screenshot with primitive:

primitive_cylinders

This is without primitive:

without_primitive_cylinders

code for the generation:
```
from fury import window, actor
import numpy as np

scene = window.Scene()
showm = window.ShowManager(scene,size=(1024, 720), reset_camera=False)
showm.initialize()

centers = np.random.rand(11,3) * 3
directions = np.random.rand(11,3) - 0.5
heights = np.random.rand(11)
colors = np.random.rand(11,3)

cylinders = actor.cylinder(centers, directions, colors, 
                           radius=0.3 ,heights=heights, capped=True,
#                         repeat_primitive=False)

scene.add(cylinders)
showm.start()
window.record(scene, out_path='viz_cylinders.png',size =(1024,720))
```

File: fury/actor.py (hunk 2 above)
Comment (Clarkszw) [reply fury/actor.py]: OK la!

**Review by skoudoro — COMMENTED** [review]
Hi @Clarkszw,

After testing your PR, this is the result I obtain: 

image


Here, we can clearly see that there is a direction issue when we are using the same direction for both methods (`repeat_primitive=True/False`).

here the updated script:
```python
from fury import window, actor
import numpy as np

scene = window.Scene()
showm = window.ShowManager(scene, size=(1024, 720), reset_camera=False)
showm.initialize()

centers = np.random.rand(11, 3) * 3
directions = np.random.rand(11, 3) - 0.5
heights = np.random.rand(11)
colors_1 = np.random.rand(11, 3)
colors_2 = np.random.rand(11, 3)

cylinders_1 = actor.cylinder(centers, directions, colors_1,
                             radius=0.3, heights=heights, capped=True,
                             repeat_primitive=False)
cylinders_2 = actor.cylinder(centers, directions, colors_2,
                             radius=0.3, heights=heights, capped=True,
                             repeat_primitive=True)

scene.add(cylinders_1, cylinders_2)
showm.start()
```

For now, we can not merge this PR, can you look at this issue and add the associated tests.

Thank you in advance

**Comment by Clarkszw** [comment]
> Hi @Clarkszw,
> 
> 
> 
> After testing your PR, this is the result I obtain: 
> 
> 
> 
> image
> 
> 
> 
> 
> 
> Here, we can clearly see that there is a direction issue when we are using the same direction for both methods (`repeat_primitive=True/False`).
> 
> 
> 
> here the updated script:
> 
> ```python
> 
> from fury import window, actor
> 
> import numpy as np
> 
> 
> 
> scene = window.Scene()
> 
> showm = window.ShowManager(scene, size=(1024, 720), reset_camera=False)
> 
> showm.initialize()
> 
> 
> 
> centers = np.random.rand(11, 3) * 3
> 
> directions = np.random.rand(11, 3) - 0.5
> 
> heights = np.random.rand(11)
> 
> colors_1 = np.random.rand(11, 3)
> 
> colors_2 = np.random.rand(11, 3)
> 
> 
> 
> cylinders_1 = actor.cylinder(centers, directions, colors_1,
> 
>                              radius=0.3, heights=heights, capped=True,
> 
>                              repeat_primitive=False)
> 
> cylinders_2 = actor.cylinder(centers, directions, colors_2,
> 
>                              radius=0.3, heights=heights, capped=True,
> 
>                              repeat_primitive=True)
> 
> 
> 
> scene.add(cylinders_1, cylinders_2)
> 
> showm.start()
> 
> ```
> 
> 
> 
> For now, we can not merge this PR, can you look at this issue and add the associated tests.
> 
> 
> 
> Thank you in advance
> 
> 

I saw your screenshot that two methods generated perpendicular cylinders(almost). But mine deviations are random..

image

I will try to check how two methods using “directions”..

**Comment by skoudoro** [comment]
In both cylinder actor, are you making sure to use the same direction (random or not) , the same direction should be used. 

If it work correctly, they should blend and  be identical, with weird color.

in your case, they should all be the same so there is clearly some wrong in the script you run.

**Comment by Clarkszw** [comment]
I have fix the `prim_cylinder()',now it create the primitive cylinder along x axis as `no primitive`.

image
Both methods generates the same shape of cylinders at the same positions.

If we shift a bit the centers of two methods, one can see they are parallel.

image


Here is code for the test:
```python
from fury import window, actor
import numpy as np

scene = window.Scene()
showm = window.ShowManager(scene, size=(1024, 720), reset_camera=False)
showm.initialize()

centers = np.random.rand(11, 3) * 3
directions = np.random.rand(11, 3) - 0.5
heights = np.random.rand(11)
colors_1 = np.random.rand(11, 3)
colors_2 = np.random.rand(11, 3)

cylinders_1 = actor.cylinder(centers, directions, colors_1,
                             radius=0.3, heights=heights, capped=True,
                             repeat_primitive=False)

centers_2 = centers.copy()
# centers_2 = centers + 0.1

cylinders_2 = actor.cylinder(centers_2, directions, colors_2,
                             radius=0.3, heights=heights, capped=True,
                             repeat_primitive=True)

scene.add(cylinders_1, cylinders_2)
showm.start()
```

**Comment by skoudoro** [comment]
Nice, thanks @Clarkszw! no time today or this weekend, so I will look more in details this Monday  or Tuesday

**Review by skoudoro — APPROVED** [review]
All good, Thank you @Clarkszw, merging

**Pull request merged**